### PR TITLE
feat: converted top-level files in src/controllers to TS good first i…

### LINKED ISSUE #672: Convert top-level files in `src/controllers` to TS
**Is your feature request related to a problem? Please describe.**
Convert top-level JS files in `src/controllers` to TS

**Describe the solution you'd like**
Convert [these top-level files](https://github.com/nkowaokwu/igbo_api/tree/master/src/controllers) (ignoring any files in utils) to TS

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Stancobridge**
## Describe your changes
This PR converts the top-levels files in `src/controllers` folder from JS to TS. I also noticed some implementations that needs modifications and tried my best to update the implementations. I also added some comments to help check some implementations if they are required

## Issue ticket number and link
Closed #672 

## How Has This Been Tested?
I ran the test with `sudo yarn test`, I used sudo because my terminal is not permitted to create the `db` folder and it returned the response below

## Screenshots (if appropriate):
igbo api test

### CODE REVIEW

**Comment by ijemmao** [comment]
@Stancobridge are you still working on this PR?

**Comment by Stancobridge** [comment]
> @Stancobridge are you still working on this PR?

Nope am done

**Comment by ijemmao** [comment]
@Stancobridge please address the failing CI build - I'm not able to merge your work until it passes all green
image

It seems like two tests are failing

## Error log
Screen Shot 2023-07-03 at 10 51 18 PM

File: src/controllers/utils/index.ts [hunk 1]
```diff
@@ -112,7 +106,7 @@ export const handleQueries = async ({
   } = query;
   const { id } = params;
   let allVerbsAndSuffixes;
-  const hasQuotes = keywordQuery && (keywordQuery.match(/["'].*["']/) !== null);
+  const hasQuotes = keywordQuery && keywordQuery.match(/["'].*["']/) !== null;
```
Comment (ijemmao) [inline src/controllers/utils/index.ts:109]: revert this change

File: src/controllers/words.ts [hunk 2]
```diff
@@ -47,14 +51,16 @@ const getWordsFromDatabase = async (req, res, next) => {
       isUsingMainKey,
       redisClient,
     } = await handleQueries(req);
+
```
Comment (ijemmao) [inline src/controllers/words.ts:54]: revert this change

File: src/controllers/words.ts [hunk 3]
```diff
@@ -1,37 +1,41 @@
...
+export const getWordData = (req: Request, res: Response, next: NextFunction) => {
   try {
     const { keyword } = req.query;
     const searchWord = removePrefix(keyword);
+
     if (!searchWord) {
       throw new Error(NO_PROVIDED_TERM);
     }
     const { wordReg: regexWord } = createRegExp(searchWord);
+
```
Comment (ijemmao) [inline src/controllers/words.ts:28]: revert this change

File: src/controllers/words.ts [hunk 4]
```diff
@@ -1,37 +1,41 @@
...
-import { handleWordFlags } from '../APIs/FlagsAPI';
-import minimizeWords from './utils/minimizeWords';
 
 /* Gets words from JSON dictionary */
-export const getWordData = (req, res, next) => {
+export const getWordData = (req: Request, res: Response, next: NextFunction) => {
   try {
     const { keyword } = req.query;
     const searchWord = removePrefix(keyword);
+
```
Comment (ijemmao) [inline src/controllers/words.ts:23]: revert this change

File: src/controllers/words.ts [hunk 5]
```diff
@@ -86,31 +96,31 @@ const getWordsFromDatabase = async (req, res, next) => {
...
-      });
+    });
+
+    if (!data.words[0]) {
+      throw new Error('No word exists with the provided id.');
+    }
+    const { words } = handleWordFlags({ data, flags });
+    const minimizedWords = minimizeWords(words, version);
+    const updatedWord = minimizedWords[0];
+
```
Comment (ijemmao) [inline src/controllers/words.ts:123]: revert this change

File: src/controllers/words.ts [hunk 6]
```diff
@@ -162,6 +163,6 @@ export const createWord = async (data, connection) => {
   /* Wait for all the Examples to be created and then add them to the Word document */
   const resolvedExamples = await Promise.all(savedExamples);
   const exampleIds = getDocumentsIds(resolvedExamples);
-  newWord.examples = exampleIds;
+  (newWord as any).examples = exampleIds;
```
Comment (ijemmao) [inline src/controllers/words.ts:166]: we don't want to use `any` please use `unknown` if you're unable to use a more specific type

**Review by ijemmao — CHANGES_REQUESTED** [review]
i appreciate this PR! but i think it's grown a bit out of scope since you've introduced your own changes after converting these files to TS and adding types

i know there's parts of the code that can be written better, but converting to TS and making changes to improve code readability (that's doesn't relate to TS, i.e. `await` vs. `.then`) makes it more challenging for me to see what changes were made only for converting from JS to TS

since i had to leave a lot of comments, I recommend starting fresh - that way you don't possibly miss a file where you didn't revert your changes

the added benefit of starting fresh is that the two failing tests will stop failing!

**Comment by Stancobridge** [comment]
Alright doing that

**Comment by ijemmao** [comment]
alright @Stancobridge i'm closing this PR so you can start a fresh one!

**Pull request closed without merging**